PLM-4 | Vijith | Expose API to validate OTP and generate token

vijithmv499 opened the pull request.

vijithmv499 committed:
PLM-4 | Vijith | Expose new endpoint to validate OTP and generate token

- Create user table with basic attributes
- Add configs required for JWT encoding to credentials file
- Add factory_bot and faker gems for test fixtures
- Add dotenv gem for configuring environment variable

vijithmv499 committed:
PLM-4 | Vijith | Refactor OtpRequest model validation specs to use FactoryBot

vijithmv499 commented on app/models/otp_request.rb:
```diff
@@ -8,4 +8,8 @@ class OtpRequest < ApplicationRecord
   validates :mobile_number, format: { with: VALID_MOBILE_NUMBER_REGEX }, length: { is: MOBILE_NUMBER_LENGTH }
+
+  def expired?
+    expire_at && expire_at < Time.now
```
This nil check is because we had missed adding non-null constraints on relevant columns of otp_requests table. I've a separate [ticket ](https://vijithmv.atlassian.net/browse/PLM-30)to fix this issue. This check will be removed as part of PLM-30

vijithmv499 committed:
PLM-4 | Vijith | Fix rubocop failures

Navinnair commented on app/controllers/authentication_controller.rb:
```diff
@@ -0,0 +1,14 @@
+      render json: { errors: [response.error] }, status: :unprocessable_entity
+    else
+      render json: { data: { token: response.token } }, status: :ok
+    end
```
I think separate variable assignments for `mobile_number` and `otp` are unnecessary here. We can directly use `params[:mobile_number]` and `params[:otp]` when initializing the AuthenticationService.

```suggestion
    service = AuthenticationService.new(params[:mobile_number], params[:otp])
    response = service.process

    if response.error.present?
      render json: { errors: [response.error] }, status: :unprocessable_entity
    else
      render json: { data: { token: response.token } }, status: :ok
    end
```

vijithmv499 replied:
done

Navinnair commented on app/services/authentication_service.rb:
```diff
@@ -0,0 +1,57 @@
+  end
+
+  def find_or_create_user!
+    User.where(mobile_number: mobile_number).first_or_create!
```
```suggestion
    User.find_or_create_by!(mobile_number: mobile_number)
```

Navinnair commented on app/services/authentication_service.rb:
```diff
@@ -0,0 +1,57 @@
+  end
+
+  def find_otp_request
+    OtpRequest.find_by(mobile_number: mobile_number, otp: otp)
```
If not already present, please consider adding a compound index on (mobile_number, otp). This table will likely grow, and a compound index could significantly improve query performance for this common lookup.

vijithmv499 replied:
Good catch 👍  Done.

Navinnair commented on spec/models/otp_request_spec.rb:
```diff
@@ -7,7 +7,7 @@
     context 'When mobile number and otp are of exactly 10 and 6 digits respectively' do
       it 'should be valid' do
-        otp = OtpRequest.new(mobile_number: valid_mobile_number, otp: valid_otp, expire_at: 10.minutes.from_now)
+        otp = build(:otp_request, mobile_number: valid_mobile_number, otp: valid_otp, expire_at: 10.minutes.from_now)
```
Optional: We could use the [shoulda-matchers](https://github.com/thoughtbot/shoulda-matchers) gem to write easier tests. It would help us check many things about our model in a simpler way. 

For example, we could write: `it { should validate_presence_of(:mobile_number) }`

vijithmv499 replied:
My initial thought process was to be very much explicit on what is being tested. But I think it makes sense to have shoulda_matchers as it will be too much to have custom specs added for every new association/validations added in the future. 

Added the gem and updated the specs

Navinnair commented on spec/models/otp_request_spec.rb:
```diff
@@ -65,7 +65,7 @@
     context 'When otp is of length greater than 6 digits' do
       it 'should not be valid' do
         invalid_otp = '300099282'
```
Let's move repeated values to `let` blocks. This will make our tests cleaner and easier to change. We can set up `valid_otp`, `invalid_mobile_number` and `expire_at` as shared values.

Also using `let` instead of creating values in each test can help with test speed (but the effect is small here). 
`let` is lazy-loaded, so it only runs when needed. This can save time if the value isn't used in every test.
But for simple values like these, the time saved is very little. The main benefit is making the code cleaner and easier to change.

vijithmv499 replied:
Specs are updated to use shoulda_matchers. Resolving this comment. Please feel free to re-open it if we have anything more to discuss here

Navinnair commented on spec/services/authentication_service_spec.rb:
```diff
@@ -0,0 +1,74 @@
+
+      context 'with no user record present in db having the given mobile number' do
+        it 'creates a user and return the JWT token generated' do
+          service = described_class.new(mobile_number, otp)
```
Optional:

 Can we use `[subject](https://github.com/rubocop/rspec-style-guide/issues/6)` instead of creating a variable. This is a common RSpec pattern and makes the test cleaner.

vijithmv499 replied:
This spec is outdated now. 

I still have a spec for expired? method. But I cannot change it to use subject as my instance is defined under the `it` block. (I can't move it to the describe / context blocks as the attributes are different for different `it` blocks)

vijithmv499 committed:
PLM-4 | Vijith | Address review comments

- Add composite index on mobile_number and otp fields in otp_requests table
- Add missing NOT_NULL constraints on otp_requests table attributes
- Rewrite otp_requests validation specs with shoulda_matchers

vijithmv499 commented:
@Navinnair Addressed the PR comments. Yet to check the audit failure. I'll check this tomorrow and let you know once that is also resolved. But please go over the other changes and see if you have any other concerns

vijithmv499 committed:
PLM-4 | Vijith | Update gem rexml to fix the security vulnerability

Ref: https://github.com/ruby/rexml/security/advisories/GHSA-4xqq-m2hx-25v8

vijithmv499 commented:
@Navinnair The bundle audit issue also have been addressed now. Please check again

Navinnair commented:
LGTM 🚀
